File: internal/util/patch.ts
import { spawn } from "node:child_process";
import { stat } from "node:fs/promises";
import path from "node:path";

export class NotGitRepoError extends Error {
  constructor() {
    super("workspace is not a git repository (.git not found)");
  }
}

export interface PatchApplyResult {
  applied: boolean;
  stdout?: string;
  stderr?: string;
}

export async function applyUnifiedDiff(
  workspace: string,
  diff: string,
  signal?: AbortSignal,
): Promise<PatchApplyResult> {
  if (!diff.trim()) {
    throw new Error("diff is empty");
  }
  const gitDir = path.join(path.resolve(workspace), ".git");
  try {
    await stat(gitDir);
  } catch {
    throw new NotGitRepoError();
  }

  return await new Promise<PatchApplyResult>((resolve, reject) => {
    const child = spawn("git", ["apply", "--whitespace=nowarn", "-"], {
      cwd: workspace,
      stdio: ["pipe", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";

    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });

    const timer = setTimeout(() => {
      child.kill("SIGKILL");
    }, 60_000);

    const abortHandler = () => {
      child.kill("SIGKILL");
    };
    signal?.addEventListener("abort", abortHandler);

    child.on("error", (err) => {
      clearTimeout(timer);
      signal?.removeEventListener("abort", abortHandler);
      reject(err);
    });

    child.on("close", (code) => {
      clearTimeout(timer);
      signal?.removeEventListener("abort", abortHandler);
      if (code === 0) {
        resolve({ applied: true, stdout, stderr });
        return;
      }
      const err = new Error(`git apply failed (exit ${code ?? 1})`);
      Object.assign(err, { result: { applied: false, stdout, stderr } });
      reject(err);
    });

    child.stdin.write(diff);
    child.stdin.end();
  });
}

File: internal/util/patch.py
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from .cancel import CancelToken


class NotGitRepoError(RuntimeError):
    def __init__(self) -> None:
        super().__init__("workspace is not a git repository (.git not found)")


@dataclass
class PatchApplyResult:
    applied: bool
    stdout: Optional[str] = None
    stderr: Optional[str] = None


def apply_unified_diff(workspace: str, diff: str, signal: CancelToken | None = None) -> PatchApplyResult:
    if not diff or not diff.strip():
        raise ValueError("diff is empty")
    git_dir = Path(workspace).resolve() / ".git"
    if not git_dir.exists():
        raise NotGitRepoError()

    process = subprocess.Popen(
        ["git", "apply", "--whitespace=nowarn", "-"],
        cwd=str(Path(workspace).resolve()),
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    def kill() -> None:
        if process.poll() is None:
            process.kill()

    if signal:
        import threading

        def watch_cancel() -> None:
            signal.wait()
            kill()

        threading.Thread(target=watch_cancel, daemon=True).start()

    try:
        stdout, stderr = process.communicate(input=diff, timeout=60)
    except subprocess.TimeoutExpired:
        kill()
        stdout, stderr = process.communicate()
        err = RuntimeError("git apply failed (timeout)")
        setattr(err, "result", PatchApplyResult(applied=False, stdout=stdout, stderr=stderr))
        raise err

    if process.returncode == 0:
        return PatchApplyResult(applied=True, stdout=stdout, stderr=stderr)

    err = RuntimeError(f"git apply failed (exit {process.returncode or 1})")
    setattr(err, "result", PatchApplyResult(applied=False, stdout=stdout, stderr=stderr))
    raise err
